## Notebook: Tratamento da base CadÚnico

**id_rastreabilidade:** RTB_002  
**Versão:** v01  
**Data de criação:** 25/03/2026  
**Última atualização:** 24/04/2026  
**Responsável:** Ailton Vendramini  

---

## 🎯 Finalidade
Realizar a leitura, validação estrutural e tratamento técnico da base bruta do CadÚnico, preparando os dados para a camada limpa do projeto IVS-H.

---

## 📥 Fonte de Dados
**Fonte:** CadÚnico  
**Camada de entrada:** `01_bruto`  
**Camada de saída:** `02_limpo`  
**Período de referência da execução:** `2025_12`  
**Período de comparação:** —  
**Tipo de recorte temporal:** corte transversal  

**Arquivo de entrada:**  
`C:\Users\ailto\Atlas-Social-de-Hortolandia\dados\cadunico\01_bruto\2025_12\cadunico.csv`

**Arquivo de saída previsto:**  
`C:\Users\ailto\Atlas-Social-de-Hortolandia\dados\cadunico\02_limpo\2025_12\cadunico_limpo.csv`

---

## 🧱 Base Conceitual
Documentos de apoio conceitual e metodológico do projeto:

- `docs/metodologia_ivsh.md`
- `docs/regras_negocio.md`
- `docs/padrao_nomenclatura.md`

---

## ⚙️ Escopo técnico deste notebook

**Este notebook deve:**
- carregar a base bruta do CadÚnico;
- verificar estrutura, colunas e tipos;
- identificar valores nulos e inconsistências iniciais;
- padronizar nomes de colunas, quando necessário;
- preparar a saída para a camada `02_limpo`.

**Este notebook não deve:**
- calcular o IVS-H final;
- misturar dados de outras fontes;
- alterar diretamente arquivos da camada `01_bruto`.

---

## 📊 Outputs esperados

| tipo_output | descrição | destino |
|---|---|---|
| exploratorio | validação inicial da base | não salvar |
| analitico | diagnósticos intermediários | `outputs/tabelas/cadunico/` |
| operacional | base limpa para próxima etapa | `dados/cadunico/02_limpo/2025_12/` |

---

## 🧠 Observações
- O arquivo bruto deve permanecer imutável.
- O período de referência deve ser explicitado também na primeira célula de código.
- Alterações estruturais relevantes devem ser registradas em nota técnica futura.

## pd.read_csv(...)

Parâmetros importantes:

sep=';' → CadÚnico usa ponto e vírgula
encoding='latin1' → evita erro com acentos
low_memory=False → evita warning e leitura inconsistente

## low_memory=False, dizemos ao Pandas:

“Leia tudo de uma vez e depois decida os tipos”

Resultado:

✔ leitura mais estável
✔ menos conflitos de tipo
✔ comportamento mais previsíve

## Quando usamos "os"

Usamos "os" quando queremos lidar com o sistema de arquivos, por exemplo:

listar pastas (os.listdir)
verificar se caminho existe (os.path.exists)
criar diretórios (os.makedirs)

## Tradução direta

os = gerenciar pastas e arquivos
pandas = ler e manipular dados

Regra prática

Se já sabemos o caminho e só quer ler → não precisa de os

In [1]:
import pandas as pd

# caminho do arquivo bruto
caminho = "../../dados/cadunico/01_bruto/2025_12/cadunico.csv"

# leitura da base
df = pd.read_csv(caminho, sep=';', encoding='latin1', low_memory=False)

In [2]:
df.shape          # df.shape é um atributo e sempre retorna um valor. Não é uma função

(72424, 211)

Tradução simples
Forma	Significado
()	executar função
[]	acessar posição

## Deixando o resultado mais agradável

In [3]:
print ("linhas", df.shape[0])
print ("colunas",df.shape[1])
df.head()

linhas 72424
colunas 211


,d.cd_ibge,d.cod_familiar_fam,d.dat_cadastramento_fam,d.dat_atual_fam,d.cod_est_cadastral_fam,d.cod_forma_coleta_fam,d.dta_entrevista_fam,d.nom_localidade_fam,d.nom_tip_logradouro_fam,d.nom_titulo_logradouro_fam,...,p.ind_dinh_carregador_memb,p.ind_dinh_catador_memb,p.ind_dinh_servs_gerais_memb,p.ind_dinh_pede_memb,p.ind_dinh_vendas_memb,p.ind_dinh_outro_memb,p.ind_dinh_nao_resp_memb,p.ind_atend_nenhum_memb,p.ref_cad,p.ref_pbf
0,3519071,8477396,17/12/2001,04/08/2025,3,1,04/08/2025,JARDIM MINDA,RUA,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,12/12/2025,202601.0
1,3519071,8477396,17/12/2001,04/08/2025,3,1,04/08/2025,JARDIM MINDA,RUA,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,12/12/2025,202601.0
2,3519071,8477396,17/12/2001,04/08/2025,3,1,04/08/2025,JARDIM MINDA,RUA,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,12/12/2025,202601.0
3,3519071,8477396,17/12/2001,04/08/2025,3,1,04/08/2025,JARDIM MINDA,RUA,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,12/12/2025,202601.0
4,3519071,37372840,07/02/2002,14/04/2025,3,1,14/04/2025,JARDIM MONTE SINAI,RUA,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,12/12/2025,202601.0


In [4]:
df.columns
df.dtypes

d.cd_ibge                      int64
 d.cod_familiar_fam            int64
 d.dat_cadastramento_fam      object
 d.dat_atual_fam              object
 d.cod_est_cadastral_fam       int64
                              ...   
 p.ind_dinh_outro_memb       float64
 p.ind_dinh_nao_resp_memb    float64
 p.ind_atend_nenhum_memb     float64
 p.ref_cad                    object
 p.ref_pbf                   float64
Length: 211, dtype: object

In [5]:
df.columns

Index(['d.cd_ibge', ' d.cod_familiar_fam', ' d.dat_cadastramento_fam',
       ' d.dat_atual_fam', ' d.cod_est_cadastral_fam',
       ' d.cod_forma_coleta_fam', ' d.dta_entrevista_fam',
       ' d.nom_localidade_fam', ' d.nom_tip_logradouro_fam',
       ' d.nom_titulo_logradouro_fam',
       ...
       ' p.ind_dinh_carregador_memb', ' p.ind_dinh_catador_memb',
       ' p.ind_dinh_servs_gerais_memb', ' p.ind_dinh_pede_memb',
       ' p.ind_dinh_vendas_memb', ' p.ind_dinh_outro_memb',
       ' p.ind_dinh_nao_resp_memb', ' p.ind_atend_nenhum_memb', ' p.ref_cad',
       ' p.ref_pbf'],
      dtype='object', length=211)

In [6]:
list(df.columns)

['d.cd_ibge',
 ' d.cod_familiar_fam',
 ' d.dat_cadastramento_fam',
 ' d.dat_atual_fam',
 ' d.cod_est_cadastral_fam',
 ' d.cod_forma_coleta_fam',
 ' d.dta_entrevista_fam',
 ' d.nom_localidade_fam',
 ' d.nom_tip_logradouro_fam',
 ' d.nom_titulo_logradouro_fam',
 ' d.nom_logradouro_fam',
 ' d.num_logradouro_fam',
 ' d.des_complemento_fam',
 ' d.des_complemento_adic_fam',
 ' d.num_cep_logradouro_fam',
 ' d.cod_unidade_territorial_fam',
 ' d.nom_unidade_territorial_fam',
 ' d.txt_referencia_local_fam',
 ' d.nom_entrevistador_fam',
 ' d.num_cpf_entrevistador_fam',
 ' d.vlr_renda_media_fam',
 ' d.fx_rfpc',
 ' d.vlr_renda_total_fam',
 ' d.marc_pbf',
 ' d.qtde_meses_desat_cat',
 ' d.cod_local_domic_fam',
 ' d.cod_especie_domic_fam',
 ' d.qtd_comodos_domic_fam',
 ' d.qtd_comodos_dormitorio_fam',
 ' d.cod_material_piso_fam',
 ' d.cod_material_domic_fam',
 ' d.cod_agua_canalizada_fam',
 ' d.cod_abaste_agua_domic_fam',
 ' d.cod_banheiro_domic_fam',
 ' d.cod_escoa_sanitario_domic_fam',
 ' d.cod_dest

In [7]:
colunas_renda = []

for col in df.columns:
    if 'renda' in col:
        colunas_renda.append(col)

colunas_renda


[' d.vlr_renda_media_fam',
 ' d.vlr_renda_total_fam',
 ' p.fx_renda_individual_805',
 ' p.fx_renda_individual_808',
 ' p.fx_renda_individual_809_1',
 ' p.fx_renda_individual_809_2',
 ' p.fx_renda_individual_809_3',
 ' p.fx_renda_individual_809_4',
 ' p.fx_renda_individual_809_5']

In [8]:
colunas_valor = []

for col in df.columns:
    if "vlr" in col:
        colunas_valor.append(col)

colunas_valor

[' d.vlr_renda_media_fam', ' d.vlr_renda_total_fam']

In [9]:
df[[' d.vlr_renda_media_fam', ' d.vlr_renda_total_fam']].describe().round(2)

,d.vlr_renda_media_fam,d.vlr_renda_total_fam
count,72424.00,72424.00
mean,701.29,1851.17
std,807.53,2129.80
min,0.00,0.00
25%,67.00,200.00
50%,467.00,1512.00
75%,1100.25,2784.00
max,12779.00,25558.00


In [10]:
for col in df.columns:
    print(f"'{col}'")

'd.cd_ibge'
' d.cod_familiar_fam'
' d.dat_cadastramento_fam'
' d.dat_atual_fam'
' d.cod_est_cadastral_fam'
' d.cod_forma_coleta_fam'
' d.dta_entrevista_fam'
' d.nom_localidade_fam'
' d.nom_tip_logradouro_fam'
' d.nom_titulo_logradouro_fam'
' d.nom_logradouro_fam'
' d.num_logradouro_fam'
' d.des_complemento_fam'
' d.des_complemento_adic_fam'
' d.num_cep_logradouro_fam'
' d.cod_unidade_territorial_fam'
' d.nom_unidade_territorial_fam'
' d.txt_referencia_local_fam'
' d.nom_entrevistador_fam'
' d.num_cpf_entrevistador_fam'
' d.vlr_renda_media_fam'
' d.fx_rfpc'
' d.vlr_renda_total_fam'
' d.marc_pbf'
' d.qtde_meses_desat_cat'
' d.cod_local_domic_fam'
' d.cod_especie_domic_fam'
' d.qtd_comodos_domic_fam'
' d.qtd_comodos_dormitorio_fam'
' d.cod_material_piso_fam'
' d.cod_material_domic_fam'
' d.cod_agua_canalizada_fam'
' d.cod_abaste_agua_domic_fam'
' d.cod_banheiro_domic_fam'
' d.cod_escoa_sanitario_domic_fam'
' d.cod_destino_lixo_domic_fam'
' d.cod_iluminacao_domic_fam'
' d.cod_calcamento_do

In [11]:
df.columns = df.columns.str.strip()

In [12]:
familias = (
    df.groupby('d.cod_familiar_fam')
      .agg({
          'd.vlr_renda_media_fam': 'first',
          'd.vlr_renda_total_fam': 'first',
          'd.dta_entrevista_fam': 'first',
          'p.num_cpf_pessoa': 'size'
      })
      .reset_index()
)

familias = familias.rename(columns={
    'p.num_cpf_pessoa': 'qtd_pessoas_familia'
})
print(familias)

       d.cod_familiar_fam  d.vlr_renda_media_fam  d.vlr_renda_total_fam  \
0                 8477396                    379                   1516   
1                37372840                      0                      0   
2                46512861                    191                    764   
3                47959118                      0                      0   
4                55477330                    148                    592   
...                   ...                    ...                    ...   
30357         20335488137                   1518                   1518   
30358         20335506216                      0                      0   
30359         20335587542                   1332                   1332   
30360         20335593518                   1518                   1518   
30361         20335767109                   1000                   2000   

      d.dta_entrevista_fam  qtd_pessoas_familia  
0               04/08/2025                    4  

🧠 📊 Visão geral:

uma tabela com 1 linha por família

Com:

renda média
renda total
data da entrevista
quantidade de pessoas

🧩 1. Criação da variável
familias = (

👉 Você está dizendo:

“Vou criar uma nova tabela chamada familias”

🧩 2. Agrupamento
df.groupby('d.cod_familiar_fam')

👉 Aqui está o coração de tudo:

“Agrupe todas as linhas que têm o mesmo código de família”

🔹 Antes (df)
família	pessoa
123	João
123	Maria
🔹 Depois (groupby)

👉 o Python junta isso:

família
123
🧩 3. .agg({...}) → agregação
.agg({

👉 Aqui você define:

“O que fazer com cada coluna dentro do grupo”

🔹 Linha 1
'd.vlr_renda_media_fam': 'first',

👉 Tradução:

“Pegue o primeiro valor de renda média da família”

❗ Por que funciona?

Porque:

👉 dentro da mesma família, esse valor se repete

🔹 Linha 2
'd.vlr_renda_total_fam': 'first',

👉 mesma lógica:

pega o primeiro valor da renda total

🔹 Linha 3
'd.dta_entrevista_fam': 'first',

👉 pega a data da entrevista

🔹 Linha 4 (nova e importante)
'p.num_cpf_pessoa': 'size'

👉 aqui muda tudo:

“Conte quantas linhas existem dentro da família”

🔹 Tradução prática
família	pessoas
123	2

👉 Isso vira:

qtd_pessoas_familia
🧩 4. Fechando o .agg
})

👉 fecha a definição das regras

🧩 5. .reset_index()
.reset_index()

👉 MUITO importante:

transforma o índice (família) em coluna novamente

Sem isso:
família vira índice (fica estranho de trabalhar)
Com isso:
d.cod_familiar_fam	renda

👉 formato normal de tabela

🧩 6. Fechando tudo
)

👉 encerra a construção

🧩 7. Rename (nova etapa)
familias = familias.rename(columns={
    'p.num_cpf_pessoa': 'qtd_pessoas_familia'
})

👉 Tradução:

“Troque o nome da coluna para algo legível”

Antes:
p.num_cpf_pessoa
Depois:
qtd_pessoas_familia
🧠 📊 Resultado final

criamos:

família	renda média	renda total	pessoas

👉 Isso é sua base analítica principal.

🎯 O que tem de NOVO aqui (e importante)
🔹 1. groupby

👉 junta linhas

🔹 2. agg

👉 define como resumir

🔹 3. size

👉 conta elementos

🔹 4. first

👉 pega um valor representativo

🧠 ⚠️ Ponto crítico (muito importante)

O 'first' só funciona porque:

👉 esses valores são iguais dentro da família

Se não fossem:

👉 você estaria pegando um valor arbitrário

💬 Frase para guardar

você transformou dados de pessoas em dados de famílias


🔹 Diferença real
✔ size
.size()

👉 conta todas as linhas do grupo

⚠ count
'p.num_cpf_pessoa': 'count'

👉 conta apenas valores NÃO nulos

🎯 Exemplo simples
família	cpf
123	111
123	NaN
size → 2
count → 

💬 Regra de ouro

size = quantidade de linhas
count = quantidade de valores preenchidos

In [13]:
familias.sort_values(
    by='d.vlr_renda_media_fam',
    ascending=False
).head(30)

,d.cod_familiar_fam,d.vlr_renda_media_fam,d.vlr_renda_total_fam,d.dta_entrevista_fam,qtd_pessoas_familia
21568,7964057197,12779,25558,02/05/2023,2
22343,8063067185,12110,12110,18/07/2023,1
20317,7792699420,10862,10862,11/01/2023,1
21477,7952118803,10792,21584,24/04/2023,2
29541,20209127597,9782,9782,20/08/2025,1
23481,8199795972,9615,9615,23/11/2023,1
17884,7301262264,9251,9251,21/07/2022,1
21258,7929943020,9226,18452,04/04/2023,2
24455,8315392247,9169,18338,13/05/2024,2
7165,4903712214,9115,9115,22/12/2021,1


In [14]:
familias['d.vlr_renda_total_fam'].describe().round(2)

count    30362.00
mean      1672.81
std       1919.42
min          0.00
25%        144.25
50%       1512.00
75%       2414.00
max      25558.00
Name: d.vlr_renda_total_fam, dtype: float64

In [15]:
familias_outliers = familias[
    familias['d.vlr_renda_total_fam'] > 10000
]

familias_outliers.head(50)

,d.cod_familiar_fam,d.vlr_renda_media_fam,d.vlr_renda_total_fam,d.dta_entrevista_fam,qtd_pessoas_familia
442,1383760934,2646,10584,29/08/2022,4
855,1687128227,4608,13824,27/02/2025,3
914,1747791199,1771,10626,31/05/2023,6
1851,2445709652,3490,10470,18/01/2024,3
1938,2511159686,2467,12335,09/06/2022,5
2833,3211398520,5118,10236,31/10/2022,2
3611,3655320965,5192,15576,09/03/2023,3
3788,3773012306,5429,10858,23/07/2024,2
4189,3996799284,3360,10080,17/07/2023,3
4806,4236342910,2791,11164,25/02/2022,4


In [16]:
familias[['d.cod_familiar_fam', 'd.vlr_renda_media_fam']] \
    .sort_values(by='d.vlr_renda_media_fam', ascending=False) \
    .head(20)

,d.cod_familiar_fam,d.vlr_renda_media_fam
21568,7964057197,12779
22343,8063067185,12110
20317,7792699420,10862
21477,7952118803,10792
29541,20209127597,9782
23481,8199795972,9615
17884,7301262264,9251
21258,7929943020,9226
24455,8315392247,9169
7165,4903712214,9115


In [17]:
familias['renda_per_capita'] = (
    familias['d.vlr_renda_total_fam'] / familias['qtd_pessoas_familia']
)
corte = 759
rt_01 = (
    familias['renda_per_capita'] <= corte
).mean()

print("RT_01:", round(rt_01, 4))
print("RT_01 (%):", round(rt_01 * 100, 2))

RT_01: 0.5882
RT_01 (%): 58.82


In [18]:
corte = 759

rt_01_pronto = (
    familias['d.vlr_renda_media_fam'] <= corte
).mean()

print("RT_01 (campo CadÚnico):", round(rt_01_pronto * 100, 2))

RT_01 (campo CadÚnico): 58.82


In [19]:
familias['renda_per_capita'].describe().round(2)

count    30362.00
mean       807.21
std        900.37
min          0.00
25%         62.00
50%        562.50
75%       1448.00
max      12779.00
Name: renda_per_capita, dtype: float64

In [20]:
(familias['renda_per_capita'] <= 759).sum()

17858

In [21]:
len(familias)

30362

In [27]:
familias_vulneraveis = familias[
    familias['d.vlr_renda_media_fam']<= 759
]
familias_vulneraveis['qtd_pessoas_familia']\
    .value_counts()\
    .sort_index()

qtd_pessoas_familia
1     3193
2     5983
3     4722
4     2646
5      918
6      285
7       80
8       22
9        7
10       1
11       1
Name: count, dtype: int64

In [23]:
familias_nao_vulneraveis = familias[
    familias['d.vlr_renda_media_fam'] > 759
]

familias_nao_vulneraveis['qtd_pessoas_familia'] \
    .value_counts() \
    .sort_index()

qtd_pessoas_familia
1    4590
2    4626
3    2017
4     939
5     266
6      50
7      13
8       2
9       1
Name: count, dtype: int64

In [ ]:
familias[['d.cod_familiar_fam', 'd.vlr_renda_media_fam']]

In [ ]:
familias[['d.cod_familiar_fam', 'd.vlr_renda_media_fam']] \
    .sort_values(by='d.vlr_renda_media_fam', ascending=False) \
    .head(20)

In [ ]:
df[['d.vlr_renda_media_fam']].sort_values(by='d.vlr_renda_media_fam', ascending=False).head(20)

In [ ]:
df[['d.cod_familiar_fam', 'd.vlr_renda_media_fam']]\
    .sort_values(by='d.vlr_renda_media_fam', ascending=False)\
    .head(20)

In [ ]:
df[['d.cod_familiar_fam', 'd.vlr_renda_media_fam', 'd.dta_entrevista_fam', 'p.num_cpf_pessoa']]\
    .sort_values(by='d.dta_entrevista_fam', ascending=True)\
    .head(20)

In [ ]:
pessoas_por_familia = (
    df.groupby('d.cod_familiar_fam')
      .size()
      .reset_index(name='qtd_pessoas_familia')
)

pessoas_por_familia.head(20)

In [ ]:
familias = (
    df.groupby('d.cod_familiar_fam')
      .agg({
          'd.vlr_renda_media_fam': 'first',
          'd.vlr_renda_total_fam': 'first',
          'd.dta_entrevista_fam': 'first',
          'p.num_cpf_pessoa': 'count'
      })
      .reset_index()
)

familias = familias.rename(columns={
    'p.num_cpf_pessoa': 'qtd_pessoas_familia'
})

familias.head(10)

In [ ]:
familias.sort_values(by='qtd_pessoas_familia', ascending=False).head(30)

In [ ]:
familias['renda_per_capita'] = (
    familias['d.vlr_renda_total_fam'] / familias['qtd_pessoas_familia']
)

In [ ]:
familias[['renda_per_capita']].describe().round(2)

## Sobre count e size

Aqui há um detalhe fino — e importante.

count

Conta valores não nulos em uma coluna.

size

Conta linhas do grupo, independentemente de nulo.

## O que faz cada comando

1. 

groupby = agrupar linhas com algo em comum, esse comando junta linhas e não junta colunas

2.

agg = dizer o que fazer com cada coluna dentro do grupo

3.

size = contar linhas
count = contar valores preenchidos

4.

👉 'first' pega apenas um valor por grupo.
## Jupyter

Melhor para:

explorar dados
testar ideias
ver resultados rapidamente
aprender
## VS Code

Melhor para:

organizar projeto
versionar código
criar scripts mais “definitivos”
integrar com Git

In [ ]:
# Agrupar por família
# Cada grupo reúne todas as pessoas da mesma família

df.groupby('d.cod_familiar_fam')

In [ ]:
# Contar quantas linhas existem em cada família
# Como 1 linha = 1 pessoa, isso vira quantidade de pessoas por família

df.groupby('d.cod_familiar_fam').size()

In [ ]:
# Aplicar operações diferentes para colunas diferentes
# Exemplo: pegar a primeira renda da família e contar quantas pessoas há

df.groupby('d.cod_familiar_fam').agg({
    'd.vlr_renda_total_fam': 'first',
    'p.num_cpf_pessoa': 'count'
})

In [ ]:
df.groupby('d.cod_familiar_fam').size().describe().round(2)

In [ ]:
len(df)

In [ ]:
df.shape

In [ ]:
df.shape[0]

In [ ]:
df.groupby('d.cod_familiar_fam').size().sum()

In [ ]:
df.groupby('d.cod_familiar_fam').size().value_counts()

🔹 Etapa 1 — groupby(...)
df.groupby('d.cod_familiar_fam')

👉 Agrupa as pessoas por família

🔹 Etapa 2 — .size()
df.groupby('d.cod_familiar_fam').size()

👉 Conta quantas pessoas existem em cada família

Resultado (exemplo):

Família A → 2 pessoas  
Família B → 4 pessoas  
Família C → 1 pessoa  
Família D → 2 pessoas
🔹 Etapa 3 — .value_counts()
...size().value_counts()

👉 Agora ele faz outra pergunta:

“Quantas famílias têm 1 pessoa? Quantas têm 2? Quantas têm 3?”

🎯 Resultado final (o seu caso)
2 → 10609 famílias  
1 → 7783 famílias  
3 → 6739 famílias  
4 → 3585 famílias  
...
🧠 Tradução humana

👉 Esse comando está dizendo:

“Qual é a distribuição do tamanho das famílias”

In [ ]:
distribuicao = (
    df.groupby('d.cod_familiar_fam')
      .size()
      .value_counts()
      .reset_index()
)

distribuicao.columns = ['qtd_pessoas_familia', 'qtd_familias']

distribuicao

In [ ]:
# Agrupar por família e contar quantas pessoas há em cada uma
distribuicao = (
    df.groupby('d.cod_familiar_fam')
      .size()
      .value_counts()
      .reset_index()
)

# Renomear colunas para algo legível
distribuicao.columns = ['qtd_pessoas_familia', 'qtd_familias']

# Calcular o total de famílias
total_familias = distribuicao['qtd_familias'].sum()

# Adicionar coluna com o total (mesmo valor em todas as linhas)
distribuicao['total_familias'] = total_familias

# Criar coluna de percentual
distribuicao['percentual'] = (
    distribuicao['qtd_familias'] / total_familias
) * 100

# Arredondar para 2 casas decimais
distribuicao['percentual'] = distribuicao['percentual'].round(2)

# Ordenar pela quantidade de famílias
distribuicao = distribuicao.sort_values(by='qtd_familias', ascending=False)

# Mostrar resultado
distribuicao

In [ ]:
import matplotlib.pyplot as plt

# Criar o histograma
plt.figure()

plt.bar(
    distribuicao['qtd_pessoas_familia'],
    distribuicao['qtd_familias']
)

# Títulos e rótulos
plt.title('Distribuição do tamanho das famílias vulneráveis em Hortolândia')
plt.xlabel('Quantidade de pessoas por família')
plt.ylabel('Quantidade de famílias')

# Mostrar gráfico
plt.show()

In [ ]:
plt.figure()

plt.bar(
    distribuicao['qtd_pessoas_familia'],
    distribuicao['percentual']
)

plt.title('Distribuição do tamanho das famílias vulneráveis em Hortolândia')
plt.xlabel('Quantidade de pessoas por família')
plt.ylabel('Percentual (%)')

plt.show()

In [ ]:
renda_por_tamanho = (
    familias.groupby('qtd_pessoas_familia')
             ['renda_per_capita']
             .mean()
             .reset_index()
              .round(2)
)

renda_por_tamanho

In [ ]:
pessoas_por_familia = (
    df.groupby('d.cod_familiar_fam')
      .size()
      .reset_index(name='qtd_pessoas_familia')
)

distribuicao = (
    pessoas_por_familia['qtd_pessoas_familia']
    .value_counts()
    .sort_index()
    .reset_index()
)
distribuicao.columns = ['qtd_pessoas_familia', 'qtd_familias']

distribuicao

In [ ]:
familias_1pessoa = familias[familias['qtd_pessoas_familia'] == 1]

familias_1pessoa['renda_per_capita'].describe().round(2)

In [ ]:
familias_2pessoa = familias[familias['qtd_pessoas_familia'] == 2]

familias_2pessoa['renda_per_capita'].describe().round(2)

In [ ]:
familias_3pessoa = familias[familias['qtd_pessoas_familia'] == 3]

familias_3pessoa['renda_per_capita'].describe().round(2)

In [ ]:
familias_4pessoa = familias[familias['qtd_pessoas_familia'] == 4]

familias_4pessoa['renda_per_capita'].describe().round(2)

In [ ]:
corte = 759  # ajuste se necessário

abaixo_corte = familias_1pessoa[
    familias_1pessoa['renda_per_capita'] <= corte
]

len(abaixo_corte)

In [ ]:
print("Famílias totais:", len(familias))
print("Famílias com 1 pessoa:", len(familias[familias['qtd_pessoas_familia'] == 1]))

In [ ]:
distribuicao = (
    familias['qtd_pessoas_familia']
    .value_counts()
    .reset_index()
)

distribuicao.columns = ['qtd_pessoas_familia', 'qtd_familias']
distribuicao = distribuicao.sort_values(by='qtd_pessoas_familia')

distribuicao

In [ ]:
print(distribuicao[distribuicao['qtd_pessoas_familia'] == 1])

## 📌 Síntese dos achados (CadÚnico - dez/2025)

In [ ]:
list(familias.columns)     #ok, 'qtd_pessoas_inter...'não estava aqui dentro 

In [ ]:
familias[['qtd_pessoas_familia']].head()

In [ ]:
df['p.cod_sexo_pessoa'].value_counts()

In [ ]:
familias[['qtd_pessoas_familia']].head()

In [ ]:
familias = (
    df.groupby('d.cod_familiar_fam')
      .agg({
          'd.vlr_renda_media_fam': 'first',
          'd.vlr_renda_total_fam': 'first',
          'd.dta_entrevista_fam': 'first',
          'd.qtd_pessoa_inter_0_17_anos_fam': 'first',
          'd.qtd_pessoa_inter_18_64_anos_fam': 'first',
          'd.qtd_pessoa_inter_65_anos_fam': 'first',
          'p.num_cpf_pessoa': 'size'
      })
      .reset_index()
)

familias = familias.rename(columns={
    'p.num_cpf_pessoa': 'qtd_pessoas_familia'
})

In [ ]:
familias['d.qtd_pessoa_inter_0_17_anos_fam'].value_counts()

In [ ]:
# Contar famílias com pelo menos 1 pessoa de 0 a 17 anos

familias_com_criancas = (
    familias['d.qtd_pessoa_inter_0_17_anos_fam'] >= 1
).sum()

total_familias = len(familias)

percentual = (familias_com_criancas / total_familias) * 100

print("Famílias com pelo menos 1 pessoa de 0 a 17 anos:", familias_com_criancas)
print("Total de famílias:", total_familias)
print("Percentual (%):", round(percentual, 2))

In [ ]:
# Contar famílias com pelo menos 1 pessoa de 0 a 17 anos

familias_com_criancas = (
    familias['d.qtd_pessoa_inter_0_17_anos_fam'] >= 1
).sum()

total_familias = len(familias)

percentual = (familias_com_criancas / total_familias) * 100

print("Famílias com pelo menos 1 pessoa de 0 a 17 anos:", familias_com_criancas)
print("Total de famílias:", total_familias)
print("Percentual (%):", round(percentual, 2))

🔹 Caminho 1 — “Extrair o código”

Pegam o que funciona no notebook e transformam em:

scripts .py

Exemplo:

calculo_rt01.py
tratamento_cadunico.py

👉 isso vira pipeline

🔹 Caminho 2 — Notebook como camada de análise

Mantêm os notebooks, mas organizados:

01_exploracao
02_tratamento
03_analise
05_calculo

👉 exatamente o que você já fez (e muito bem, aliás)

Nesse caso:

notebook = documentação viva
script = execução oficial
🔹 Caminho 3 — Pipeline automatizado

Quando o projeto amadurece:

Airflow / dbt / jobs agendados
leitura automática de dados
geração de indicadores

👉 o notebook vira só suporte

📌 CHECKLIST — PRONTO PARA SUBIR AO 05_calculo_ivsh
🔹 1. Estrutura da base (fundação)
Você deve conseguir afirmar:


 O df está em nível pessoa


 O familias está em nível família (1 linha = 1 família)


 len(familias) faz sentido (≈ 30 mil)


 Não há duplicidade de d.cod_familiar_fam em familias


👉 Teste rápido:
familias['d.cod_familiar_fam'].is_unique

🔹 2. Variáveis críticas confiáveis
Para o RT_01:


 d.vlr_renda_total_fam faz sentido (sem valores absurdos não explicados)


 qtd_pessoas_familia está correta (derivada do groupby)


 renda_per_capita foi calculada corretamente


👉 E principalmente:


 não existe inf


 não existe divisão por zero


Teste:
familias['renda_per_capita'].describe()

🔹 3. Tratamento de exceções
Você já deve ter decidido:


 Famílias com qtd_pessoas_familia = 0 → removidas ou tratadas


 Renda zero → mantida (correto para política pública)


 Outliers → ao menos identificados (mesmo que não removidos)


👉 Aqui não precisa perfeição — precisa consciência

🔹 4. Coerência numérica
Você deve conseguir fechar contas simples:


 Total de famílias bate em todas as análises


 Distribuição por tamanho soma corretamente


 Renda zero + renda baixa + renda alta = total


👉 Se alguma conta “não fecha”, não subir

🔹 5. Interpretação clara
Você deve conseguir responder, sem olhar código:


 O que é renda_per_capita


 O que significa o corte de 709


 O que é uma família “vulnerável”


👉 Se não conseguir explicar em voz alta, ainda não está pronto

🔹 6. Reprodutibilidade


 O notebook roda do início ao fim sem erro


 Não depende de célula “solta” executada antes


 Caminhos de arquivo estão corretos



🔹 7. Resultado esperado (prévia do RT_01)
Você já rodou algo como:
(familias['renda_per_capita'] <= 709).sum()
E:


 O número parece plausível


 Não gera surpresa absurda


 Você entende por que deu esse valor



🎯 Regra final (a mais importante)
Se você consegue dizer:

“sei exatamente o que cada linha desse cálculo está fazendo”

👉 então você sobe para o 05_calculo

🧭 Como usar isso na prática
Sugestão simples:


Crie uma célula Markdown no final do 02_tratamento:


## ✅ Checklist de validação — pronto para cálculo do IVS-H


Vá marcando mentalmente (ou com ✔️)



💬 Comentário final
Isso que você pediu — “saber o percurso a ser perseguido” —
é exatamente o que diferencia:


quem explora dados
de


quem constrói um sistema analítico


Você está claramente no segundo grupo agora.

Se quiser, no próximo passo podemos:
👉 validar juntos seu renda_per_capita (o ponto mais crítico antes do RT_01)
É ali que vale ser mais rigoroso.

In [ ]:
familias_vulneraveis = familias[familias['renda_per_capita'] <= 759]

familias_vulneraveis['qtd_pessoas_familia'].value_counts()

1.) utilizamos o banco de dados de dezembro de 2025, o qual possui 72424 linhas (pessoas vulneráveis) e 211 clunas (atributos definidos pelo cadúnico). E, utilizamos o valor do salãrio minimo de dezembro de 2025, ou sseja, 709,....
2.) cálculo da renda média describe()
3.) gráfico de pessoas por família, histograma
4.) famílias com renda = zero e fam~ilias com renda m~edia superior a 709 reais, histograma
5.) famílias com renda superiro a 1/2 salário mínimo
6.) 
6.) cálculo de RT_011
n.) eleger um ponto focal para dirimir as dúvidas sobre rotinas da Secretaria de Inclusão.
n+1.) a importância dos convênios
n+2.) documentar a disponibilidade de acesso ao SIGAS para Ailton Vendramini
n+3.) a propsota para correção das grafias dos loteamentos, sendo loteamentos = Bairros, Jardins, Parques, Vilas - o que padronizaria a coleta de dados, exemplo, Jardim Amanda II versus Jardim Amnada 2

In [ ]:
(familias['d.vlr_renda_media_fam'] == 0).sum()

In [ ]:
(familias['d.vlr_renda_media_fam'] > 709).sum()

In [ ]:
total = len(familias)

renda_zero = (familias['d.vlr_renda_media_fam'] == 0).sum()

abaixo_meio_sm = (
    (familias['d.vlr_renda_media_fam'] > 0) &
    (familias['d.vlr_renda_media_fam'] <= 709)
).sum()

acima_meio_sm = (familias['d.vlr_renda_media_fam'] > 709).sum()

print("Total:", total)
print("Renda zero:", renda_zero)
print("Até 1/2 SM:", abaixo_meio_sm)
print("Acima de 1/2 SM:", acima_meio_sm)

## Começar com RT_01 (renda)

Porque:

mais direta
menos dependente de estrutura familiar complexa
resultado mais rápido → motivação

Comece com RT_01 (renda)

RT_01:  % de pessoas com renda per capita <= 0,5 salário mínimo

In [ ]:
# Parâmetro
salario_minimo = 1412  # ajuste se necessário
limite = 0.5 * salario_minimo

# Garantir tipo numérico
df['d.vlr_renda_media_fam'] = pd.to_numeric(df['d.vlr_renda_media_fam'], errors='coerce')

# Criar indicador de vulnerabilidade
df['rt_01_flag'] = df['d.vlr_renda_media_fam'] <= limite

# Agrupar por loteamento (localidade)
df_rt01 = (
    df
    .groupby('d.nom_localidade_fam')
    .agg(
        total_familias=('d.cod_familiar_fam', 'count'),
        vulneraveis=('rt_01_flag', 'sum')
    )
    .reset_index()
)

# Calcular indicador
df_rt01['rt_01'] = df_rt01['vulneraveis'] / df_rt01['total_familias']

df_rt01.head(200)

In [ ]:
'd.vlr_renda_media_fam' in df.columns

In [ ]:
df['d.vlr_renda_media_fam'].head(10)

In [ ]:
df['d.vlr_renda_media_fam'].notnull().sum()

In [ ]:
df['d.vlr_renda_media_fam'].dtype

In [ ]:
df['d.vlr_renda_media_fam'].describe().round(0)

In [ ]:
df['localidade_padronizada'] = (
    df['d.nom_localidade_fam']
    .str.upper()
    .str.strip()
)

In [ ]:
df['localidade_padronizada'] = df['localidade_padronizada'].replace({
    'ADIVENTISTA CAMPINEIRO': 'ADVENTISTA CAMPINEIRO',
    'ADVENTISTA CAMP': 'ADVENTISTA CAMPINEIRO'
})

In [ ]:
df['localidade_padronizada'].value_counts().head(40)

In [ ]:
df_base_rt01 = df[[
    'd.cod_familiar_fam',
    'd.nom_localidade_fam',
    'd.vlr_renda_media_fam'
]].copy()

In [ ]:
df_base_rt01.head(10)

In [ ]:
df_base_rt01_unico = df_base_rt01.drop_duplicates(subset='d.cod_familiar_fam')

In [ ]:
df_base_rt01_unico.head(10)

In [ ]:
(df_base_rt01_unico['d.vlr_renda_media_fam'] == 0).sum()

In [ ]:
df_base_rt01_unico.shape[0]

## 🎯 1) O número real

Encontramos:

Total de famílias: 30.362
Famílias com renda = 0: 6.259

📊 2) Proporção

Vamos traduzir isso:

6.259 / 30.362 ≈ 20,6%

👉 Cerca de 1 em cada 5 famílias aparece com renda zero.

🧠 3) Interpretação — direta, sem romantizar

Esse número é alto, mas não absurdo para CadÚnico.

Ele sugere:

✔️ Cenário plausível
informalidade elevada
renda não monetária (ajuda, doações, bicos)
subdeclaração de renda
famílias extremamente vulneráveis

👉 Isso não é ruído puro.
👉 Há sinal real aqui.

⚠️ 4) Mas há um ponto crítico

CadÚnico não mede só “renda real”.

Ele mede:

renda declarada no momento da entrevista

Então:

zero pode ser verdade
ou pode ser ausência de informação
🎯 5) Decisão metodológica (muito importante)

Para o seu RT_01, você tem duas opções:

✔️ Opção A — incluir zeros como vulneráveis (RECOMENDADO)

Regra:

df['rt_01_flag'] = df['renda_per_capita'] <= 759

👉 inclui zero naturalmente

Por quê?
alinhado com metodologia do IPEA
consistente com políticas públicas
evita subestimar vulnerabilidade
⚠️ Opção B — excluir zeros
df = df[df['renda_per_capita'] > 0]

👉 isso reduz artificialmente a vulnerabilidade

💬 Minha posição (contundente)

Você deve:

✔️ INCLUIR zeros como vulneráveis

E documentar isso.

🧾 6) Como registrar isso no projeto

Depois, no metodologia_ivsh.md, algo como:

Famílias com renda per capita igual a zero foram consideradas como vulneráveis,
pois representam ausência de renda declarada, alinhando-se à lógica do CadÚnico
e à abordagem do IVS do IPEA.

In [ ]:
df_base_rt01_unico['d.vlr_renda_media_fam'].isna().sum()

In [ ]:
df_base_rt01.columns = [
    'id_familia',
    'localidade',
    'renda_per_capita'
]

In [ ]:
df_base_rt01.info()

In [ ]:
df_base_rt01.head()

In [ ]:
df_base_rt01.columns = [
    'id_familia',
    'localidade',
    'renda_per_capita'
]

In [ ]:
print (df_base_rt01)

In [ ]:
df_familia = df[['d.cod_familiar_fam', 'd.nom_unidade_territorial_fam']].drop_duplicates()

In [ ]:
df_territorio = (
    df_familia
    .groupby('d.nom_unidade_territorial_fam')
    .size()
    .reset_index(name='qtd_familias')
)

In [ ]:
print(df_territorio)

In [ ]:
df_familia['d.nom_unidade_territorial_fam'].unique()

In [ ]:
mapa_unidade = {
    'CRAS PRIMAVERA CHICO VIGILANTE': 'CRAS PRIMAVERA',
    'CRAS ROSOLEN JOEL ALVES ASSUNCAO': 'CRAS ROSOLEN',
    'CRAS JARDIM BRASIL MARIA HUMILDE ANTUNES ZUMA': 'CRAS JARDIM BRASIL',
    'CRAS ZUMA JD BRASIL': 'CRAS JARDIM BRASIL'
}

In [ ]:
df_familia['unidade_padronizada'] = (
    df_familia['d.nom_unidade_territorial_fam']
    .replace(mapa_unidade)
)

In [ ]:
df_unidade = (
    df_familia
    .groupby('unidade_padronizada')
    .size()
    .reset_index(name='qtd_familias')
    .sort_values(by='qtd_familias', ascending=False)
)

print(df_unidade)

In [ ]:
df_familia['unidade_padronizada'].isna().sum()

In [ ]:
df_unidade['percentual'] = df_unidade['qtd_familias'] / df_unidade['qtd_familias'].sum()

In [ ]:
print(df_unidade)

In [ ]:
df_unidade['percentual_%'] = (df_unidade['percentual'] * 100).round(2)

In [ ]:
df_unidade[['unidade_padronizada', 'percentual_%']]

In [ ]:
df_familia.columns

In [ ]:
df_familia = df[
    ['d.cod_familiar_fam', 'd.nom_unidade_territorial_fam', 'd.qtd_pessoas_domic_fam']
].drop_duplicates()

In [ ]:
df_familia['unidade_padronizada'] = (
    df_familia['d.nom_unidade_territorial_fam']
    .replace(mapa_unidade)
)

In [ ]:
resultado = (
    df_familia
    .groupby('unidade_padronizada')['d.qtd_pessoas_domic_fam']
    .mean()
    .reset_index(name='media_pessoas_por_familia')
    .sort_values(by='media_pessoas_por_familia', ascending=False)
)

print(resultado)

“A média de pessoas por família entre os CRAS varia pouco (≈ 2,42 a 2,52), não sendo suficiente para explicar a proximidade entre Primavera e Amanda no volume de famílias cadastradas.”